### DISCUSSION TITLE
How should heap allocation when referencing be done?

### DISCUSSION THREAD

**Discussion by crthpl** [vlang/v#15065] [Ideas]
```v
a := 3
b := &a
```
becomes
```c
int *a = HEAP(int, (3));
int* b = &(*(a));
```
`a` is made `HEAP()` in `Checker.mark_as_referenced`, by setting `Var.is_auto_heap` of `a`, so that `a` is put on the heap at initialization, implicitly.
This is in case `b` is returned through a `mut` argument or returning, causing a dangling pointer.
`ast.Type` isn't `.ref()`'d, so `is_auto_heap` requires many special cases everywhere to dereference.
My first suggestion for this is to simply reference the stack object, because we know `a` isn't escaping `abc()` - its simply being assigned to a local variable `b`:
```c
int a = 3;
int* b = &a;
```
If `b` is passed as a function argument, returned, or assigned to a struct field (which is too difficult to track now, but could be changed in the future when we have a better autofree framework that tracks where values go even in struct fields), then it will simply be heap allocated with `memdup`.

This is currently an error:
```v
fn abc(c &int) &int {
	return c
}
```
because you are returning `c`, and if `c` is on the stack of the calling function and that function returns `a`, it will be a dangling pointer:
```v
a := def() // `a` would be a dangling pointer to `b`
           // `b := 3` (stack allocated) -> `abc(&b)` -> return `&b` from `abc()` -> return `&b` from `def()` -> `a`, `a == &b`
fn def() &int {
	b := 3
	return abc(&b)
}
fn abc(c &int) &int {
	return c
}
```
In the 1st example, the (current) checker marked `b` with `is_auto_heap` when it was referenced, and then gen put it on the heap. This is not possible with functions, as it requires a pass to check which reference arguments are returned from `abc()` (as the same reference, through returning or `mut` arguments), which requires a pass over all functions that `abc()` calls, etc. This requires many passes, and an additional one to mark variables who are referenced entering functions who return the same reference with `is_auto_heap`. However, the checker is single pass. This is why it is simply an error, and you can use `[heap]` to guarantee that `b` is on the heap and will not be dangling.
I don't know why this wasn't made consistent with the first example and marks `is_auto_heap` on `b` when it is referenced, regardless of where it goes.

My suggestion is to change the behaviour of calling function the same as with my first suggestion: `b` escapes `def()` through calling `abc()`, so `abc(&b)` becomes `main__abc(HEAP(int, b))`.

This would make the only use of `[heap]` to keep things always in the same memory location (e.g. lots of things in `sync`) or because the struct is very large.

**Comment by medvednikov** [comment]
> My suggestion is to change the behaviour of calling function the same as with my first suggestion

Sounds good to me.

**Comment by crthpl** [comment]
What should be done for wanting to get the address of a value? In Current V `&a` would just become `&a` in C, whereas with my proposal it would become `HEAP(a)`. This would usually be what is desired, though in unsafe code this may not be wanted, for example in the destination argument of `memcpy`, or for performance when one is 100% sure that the reference will not escape the current scope. My idea is to add a hard coded unsafe compiler function `__addr(a)` or `ADDR(a)` that would simply become `&a`.

**Reply by spytheman** [reply]
> My idea is to add a hard coded unsafe compiler function `__addr(a)` or `ADDR(a)` that would simply become `&a`.

What would taking the address of an array element or a struct field look like with __addr() ?

**Reply by medvednikov** [reply]
I think `&` should just copy the behavior of Go.

**Reply by medvednikov** [reply]
The compiler should be able to figure out whether the value escapes and do necessary operations behind the scene.

**Reply by crthpl** [reply]
> I think `&` should just copy the behavior of Go.

Go does escape analysis, and puts objects on the heap accordingly. To do escapingly, one needs to traverse through functions, starting at main.main, so we could reuse markused for that. This will increase bulid times that weren't using skip-unused by ~4%, though the skip-unused parts will be removed and escape analysis will be added, which isn't that simple as you need to see where an object goes (through structs, variables, etc.).

However, `__addr` will still be needed just like go needs `unsafe.Pointer()`

> > My idea is to add a hard coded unsafe compiler function `__addr(a)` or `ADDR(a)` that would simply become `&a`.
> 
> What would taking the address of an array element or a struct field look like with __addr() ?

Arrays would need a special case: `arr.data+index*arr.element_size`. Struct fields simple be `&a.b`

**Reply by medvednikov** [reply]
@crthpl the extra pass is fine, it'll be about 25-40ms, it's better than complicating the language.

Adding it for stuff like Go's `unsafe.Pointer()` is fine.
